packages/react-jsx-babel-plugin/src/TransformJSXToReactBabelPlugin.js

```diff
@@ -0,0 +1,586 @@
+      if (t.isJSXAttribute(attr) && t.isJSXIdentifier(attr.name)) {
+        if (attr.name.name === 'key') {
+          key = convertAttribute(attr).value;
+        } else if (attr.name.name === '__source') {
+          source = convertAttribute(attr).value;
+        } else if (attr.name.name === '__self') {
+          self = convertAttribute(attr).value;
+        } else {
+          attribs.push(attr);
+        }
+      } else {
+        attribs.push(attr);
+      }
+    }
+
+    if (attribs.length || path.node.children.length) {
+      attribs = buildJSXOpeningElementAttributes(
+        attribs,
+        file,
+        path.node.children,
+      );
+    } else {
+      // attributes should never be null
+      attribs = t.objectExpression([]);
+    }
+
+    args.push(attribs);
+
+    if (!file.opts.development) {
+      if (key !== undefined) {
+        args.push(key);
+      }
+    } else {
+      // isStaticChildren, __source, and __self are only used in development
+      args.push(
+        key === undefined ? t.identifier('undefined') : key,
+        t.booleanLiteral(path.node.children.length > 1),
+        source === undefined ? t.identifier('undefined') : source,
+        self === undefined ? t.identifier('undefined') : self,
+      );
+    }
+
+    if (opts.post) {
+      opts.post(state, file);
+    }
+    return (
+      state.call ||
+      t.callExpression(
+        path.node.children.length > 1 ? state.staticCallee : state.callee,
+        args,
+      )
+    );
+  }
+
+  // Builds props for React.jsx. This function adds children into the props
+  // and ensures that props is always an object
+  function buildJSXOpeningElementAttributes(attribs, file, children) {
+    let _props = [];
+    const objs = [];
+
+    const useBuiltIns = file.opts.useBuiltIns || false;
+    if (typeof useBuiltIns !== 'boolean') {
+      throw new Error(
+        'transform-react-jsx currently only accepts a boolean option for ' +
+          'useBuiltIns (defaults to false)',
+      );
+    }
+
+    while (attribs.length) {
+      const prop = attribs.shift();
+      if (t.isJSXSpreadAttribute(prop)) {
+        _props = pushProps(_props, objs);
+        objs.push(prop.argument);
+      } else {
+        _props.push(convertAttribute(prop));
+      }
+    }
+
+    // In React.JSX, children is no longer a separate argument, but passed in
+    // through the argument object
+    if (children && children.length > 0) {
+      if (children.length === 1) {
+        _props.push(t.objectProperty(t.identifier('children'), children[0]));
+      } else {
+        _props.push(
+          t.objectProperty(
+            t.identifier('children'),
+            t.arrayExpression(children),
+          ),
+        );
+      }
+    }
+
+    pushProps(_props, objs);
+
+    if (objs.length === 1) {
+      // only one object
+      if (!t.isObjectExpression(objs[0])) {
+        // this could be null, and jsx expects props to be non-null
```

Comment:
I think the comment is incorrect. It's either an `ObjectExpression` or the argument of a `JSXSpreadAttribute`.

Reply:
In the case of:
```
const props = null;
<div {...props} />
```
this would translate into `jsx(div, props, ...)`, which would equal `jsx(div, null, ...)`. Therefore, we have to transform the above to `jsx(div, Object.assign({}, props),...)` to ensure that the second argument is never null. You make a good point though that this comment is slightly confusing.